File: leetsync/1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.go
func numberOfSubstrings(s string) int {
	n := len(s)
	charToFreqMap := make([]int, 3)
	left, right, count := 0, 0, 0
	for right < n {
		ch := s[right]
		charToFreqMap[ch-'a']++
		for charToFreqMap[0] > 0 && charToFreqMap[1] > 0 && charToFreqMap[2] > 0 {
			count += (n - right)
			charToFreqMap[s[left]-'a']--
			left++
		}
		right++
	}
	return count
}

File: leetsync/1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.java
class Solution {
    public int numberOfSubstrings(String s) {
        int n = s.length();
        int[] charToFreqMap = new int[3];
        int left = 0, right = 0, count = 0;
        while (right < n) {
            char ch = s.charAt(right);
            charToFreqMap[ch - 'a']++;
            while (charToFreqMap[0] > 0 && charToFreqMap[1] > 0 && charToFreqMap[2] > 0) {
                count += (n - right);
                charToFreqMap[s.charAt(left) - 'a']--;
                left++;
            }
            right++;
        }
        return count;
    }
}